Custom help command.

AadilVarsh opened the pull request:
### Implemented a custom help command, which returns responses with embeds.

![tyrant-help](https://user-images.githubusercontent.com/72086030/143666596-8d60bc13-3505-4897-a290-31cf850a3306.gif)

Bluenix2 commented:
First of all, this looks great!

That said discord.py has built-in ways to create help commands and there's just a few things you need to change to make it work here.

I don't think the actual documentation covers it that nicely, but [I have it implemented in a very simple manor in a bot of mine](https://github.com/Bluenix2/PWBot/blob/master/cogs/meta.py#L10-75). All you need to do is define those same methods, and paste in the code from `create_help_embed()` that you added (but use `self.get_destination()` to get the channel to send the help to, as I do in that code).

Afterwards you need to register it [inside the `__init__()`](https://github.com/Bluenix2/PWBot/blob/master/bot.py#L35-40). The `command_attrs` is used so that the help command itself gets a brief and description here (you can skip that step if you'd like).

For the code that sends the help embeds, just change it to use [`ctx.send_help()`](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#discord.ext.commands.Context.send_help).

AadilVarsh commented:
@Bluenix2 Hey!

Thanks for the reccomendation, I quickly parsed through the code sample that you had tagged. Yup. That seems great! I'll try to implement it soon.

Bluenix2 commented on tyrant/bot.py:
```diff
@@ -81,21 +83,18 @@ async def send_connection_log(self) -> None:
         try:
             webhook = await self.fetch_webhook(constants.Webhooks.logs)
         except discord.HTTPException as e:
-            logger.error("Failed to fetch webhook to send connection log: status {}", e.status)
+            logger.error(
+                "Failed to fetch webhook to send connection log: status {}", e.status
+            )
```
Can you undo these changes presumably done by the black linter?

Bluenix2 commented on tyrant/bot.py:
```diff
@@ -81,21 +83,18 @@ async def send_connection_log(self) -> None:
         try:
             webhook = await self.fetch_webhook(constants.Webhooks.logs)
         except discord.HTTPException as e:
-            logger.error("Failed to fetch webhook to send connection log: status {}", e.status)
+            logger.error(
+                "Failed to fetch webhook to send connection log: status {}", e.status
+            )
         else:
             image_url = await github.get_random_connection_image(self.http_session)
-            embed = discord.Embed(colour=discord.Colour.dark_magenta(),)
+            embed = discord.Embed(
+                colour=discord.Colour.dark_magenta(),
+            )
```
This one as well.

Bluenix2 commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,59 @@
+        super().__init__(*args, **kwargs)
+
+    async def send_bot_help(self, mapping):
+        bot: Bot = self.context.bot
```
What's the point of this annotation?

AadilVarsh replied:
I added it for autocompletion, but now that I think of it I'll remove that as well!

Bluenix2 commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,57 @@
+    async def send_command_help(self, command: commands.Command):
+        bot: Bot = self.context.bot
+
+        for cog_name in bot.cogs:
+            cog = bot.cogs[cog_name]
+            for cmd in cog.walk_commands():
+                if cmd.name == command.name.strip().lower():
+                    help_embed = Embed(
+                        title=f"{constants.Bot.prefix}{cmd.name} [{' | '.join(list(cmd.aliases))}]",
+                        description=(
+                            cmd.help if cmd.description == "" else cmd.description
+                        ),
+                        color=constants.Bot.embed_color,
+                    )
+                break
```
I don't understand, don't you have the command right there as a parameter (called `command`)?

AadilVarsh replied:
after effects of copypasting!!! fixing it brb.

Bluenix2 commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,57 @@
+
+    async def send_error_message(self, error):
+        help_embed: Embed = Embed(
+            title="Oops!", description=error, color=constants.Bot.embed_color
```
Calling `str()` on an error doesn't actually produce a traceback. What you want to do here is use [`traceback.format_exception`](https://docs.python.org/3/library/traceback.html#traceback.format_exception) I think.

The code would look like this:
```python
description = ''.join(traceback.format_exception(type(error), error, error.__traceback__))
```

AadilVarsh replied:
I havent called str() on the error, it actually returns error as a string.

![image](https://user-images.githubusercontent.com/72086030/144706904-050d9564-799d-4dfb-92b5-4881d858000a.png)

Bluenix2 replied:
Discord.py will call `str(description)`

lemonsaurus replied:
yeah you do probably want the traceback here.

Bluenix2 commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,57 @@
+class TyrantHelp(commands.HelpCommand):
+    """Help command for tyrant."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Note: I know this is in my code but it is not required, do feel free to remove this.

lemonsaurus reviewed (changes requested):
This is pretty good, but there's a few more things to address.

Bluenix2 commented on .gitignore:
```diff
@@ -127,4 +127,7 @@ dmypy.json
+
+#Editors
+.idea
+.vscode
```
Can you just add this newline here?

Bluenix2 commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,53 @@
+class TyrantHelp(commands.HelpCommand):
+    """Help command for tyrant."""
+
+    def __init__(self):
+        super().__init__()
```
This can be removed

Bluenix2 commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,53 @@
+            cog = bot.cogs[cog_name]
+            field_value = ""
+
+            for cmd in cog.walk_commands():
+                field_value += f"**{constants.Bot.prefix}{cmd.name}** [{' | '.join(list(cmd.aliases))}]\n"
```
This doesn't handle subcommands that well. Instead of `.purge all` this makes it `.all`.

AadilVarsh commented:
I've reworked the embed styling of all of the help commands, as mentioned by @Bluenix2  in discord, about the `get_command_signature()`, I've implemented that all along the help commands. 

I've made the command/cog description, character limit to 80. Because if exceeded the entire embed looks shabby, Keeping command help strings short will be ideal for it to be rendered in the embeds. But in Individual command help embeds it will be rendered irrespective of the character length.

# New Embed Style preview.

## Bot Help embed.

![image](https://user-images.githubusercontent.com/72086030/150341317-45bb557a-438a-4b45-8a82-75b846b24dec.png)

> ### You can see there is no description for any of the purge group or its sub commands, this is due to its length exceeding 80 characters.

![image](https://user-images.githubusercontent.com/72086030/150343058-661344f8-5c79-486a-8dcf-eb82c189d3e8.png)

> ### **Commands are also filtered based on user's permissions, Admin POV 👇🏻**

![image](https://user-images.githubusercontent.com/72086030/150341347-1e629722-fc8b-4bc8-b0b4-7957af093bb6.png)

> ### In individual help commands, you can see the description being rendered irrespective of its character length.
![image](https://user-images.githubusercontent.com/72086030/150341803-58a58da6-d624-4497-a27d-c5ab6490139b.png)

Bluenix2 commented:
This looks amazing, I'll give it a last test later today!

Bluenix2 reviewed (approved):
This is great! Thank you for being patient with me here 😅.

AadilVarsh commented:
> This is great! Thank you for being patient with me here 😅.

Haha, It's actually the other way around. Thank you for the reviews!

lemonsaurus commented on tyrant/utils/help.py:
````diff
@@ -0,0 +1,102 @@
+                if cog and (cog.description and (len(cog.description) <= 80))
+                else ""
+            )
+            cmd_list = "```\n"
+
+            for cmd in cog_commands:
````
We can afford to write the word `command` instead of just `cmd`. There's really no advantage to this sort of shorthand, it just looks lazy.

lemonsaurus commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,102 @@
+
+    async def send_group_help(self, group: commands.Group):
+        """Post help for command groups."""
+
```
No newlines after coroutine docstrings.

```suggestion
```

lemonsaurus commented on tyrant/utils/help.py:
````diff
@@ -0,0 +1,102 @@
+    async def send_group_help(self, group: commands.Group):
+        """Post help for command groups."""
+
+        group_commands = await self.filter_commands(group.commands, sort=True)
+
+        if len(group_commands) == 0:
+            await self.send_command_help(group)
+            return
+
+        help_embed = Embed(title=f"{group.name.title()} {self.fmt_command_aliases(group, add_parenthesis=True)}", color=constants.Color.yellow)
+
+        msg = (group.help if group.description == "" else group.description)
+        msg = msg if len(msg) <= 80 else ''
+
+        for command in group_commands:
+            description = (
+                command.help if command.description == "" else command.description
+            ) + "\n"
+            cmd = f"```\n{self.get_command_signature(command)}\n```"
+
+            help_embed.add_field(name=command.name.title(), value=f"{description if len(description) <= 80 else ''}{cmd}", inline=False)
+
+        await self.get_destination().send(embed=help_embed)
````
This coroutine needs more _block comments_. Someone who is skimming through this file should be able to understand roughly what you're doing without reading your source code, just from the block comments. A block comment every 5-10 lines is a good rule of thumb. Obviously don't just _arbitrarily_ make block comments, make them for a logical block of code that accomplishes one specific thing.

lemonsaurus commented on tyrant/utils/help.py:
````diff
@@ -0,0 +1,102 @@
+            ) + "\n"
+            cmd = f"```\n{self.get_command_signature(command)}\n```"
+
+            help_embed.add_field(name=command.name.title(), value=f"{description if len(description) <= 80 else ''}{cmd}", inline=False)
````
This line is way too dense, too much cognitive complexity. Please refactor.

lemonsaurus commented on tyrant/utils/help.py:
```diff
@@ -0,0 +1,102 @@
+
+        help_embed = Embed(title=f"{group.name.title()} {self.fmt_command_aliases(group, add_parenthesis=True)}", color=constants.Color.yellow)
+
+        msg = (group.help if group.description == "" else group.description)
```
`message`

lemonsaurus commented:
@AadilVarsh Also, when you address review comments, would you mind resolving the comments in this thread? That way, we can see which ones you've already addressed.

AadilVarsh commented:
> @AadilVarsh Also, when you address review comments, would you mind resolving the comments in this thread? That way, we can see which ones you've already addressed.

I'm sorry, I didn't see this comment. I've resolved all of the requested changes.

lemonsaurus reviewed (approved):
Alright, let's do this.
